mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp

```diff
@@ -0,0 +1,2568 @@
+      SmallVector<int64_t> shape(
+          ShapedType::kDynamic,
+          cast<RankedTensorType>(op.getResult().getType()).getRank());
+      input = rewriter.createOrFold<tensorrt::ReshapeOp>(
+          op.getLoc(),
+          RankedTensorType::get(
+              shape, cast<RankedTensorType>(input.getType()).getElementType()),
+          input, op.getDynamicReshape());
+    }
+    input = rewriter.createOrFold<tensorrt::TransposeOp>(
+        op.getLoc(), input,
+        AffineMap::getPermutationMap(op.getSecondTranspose(), op.getContext()));
+    rewriter.replaceOp(op, input);
+    return success();
+  }
+};
+} // namespace
+
+namespace {
+template <typename OpType>
+class RankChangeToReshape : public OpRewritePattern<OpType> {
+public:
+  using OpRewritePattern<OpType>::OpRewritePattern;
+  LogicalResult matchAndRewrite(OpType op,
+                                PatternRewriter &rewriter) const override {
+    rewriter.replaceOpWithNewOp<tensorrt::ReshapeOp>(op, op.getType(),
+                                                     op.getInput());
+    return success();
+  }
+};
+} // namespace
+
+namespace {
+
+// Push down transpose to into an einsum, rearranging the axes of the input
+// tensors in the einsum as needed einsum(x1, transpose(x2), ...) -> einsum(x1,
+// x2, ...)
+class PushDownTransposeToEinsum : public OpRewritePattern<tensorrt::EinsumOp> {
+public:
+  using OpRewritePattern<tensorrt::EinsumOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensorrt::EinsumOp op,
+                                PatternRewriter &rewriter) const override {
+    std::string einsumEquation{op.getEquation()};
+    size_t pos = einsumEquation.find("->");
+    if (pos == std::string::npos)
+      return failure();
+    std::string einsumEquationLhs = einsumEquation.substr(0, pos);
+    std::string einsumEquationRhs = einsumEquation.substr(pos + 2);
+    SmallVector<std::string> einsumEquationLhsParts;
+    std::string currentPart;
+    std::istringstream lhsStream(einsumEquationLhs);
+    while (std::getline(lhsStream, currentPart, ',')) {
+      einsumEquationLhsParts.push_back(currentPart);
+    }
+
+    bool hasTransposeInput = false;
+    // Question: in the case that a transpose is used by multiple ops, might not
+    // want to push down into einsum
+    for (auto input : op.getInputs()) {
+      if (input.getDefiningOp<tensorrt::TransposeOp>()) {
+        auto perm =
+            input.getDefiningOp<tensorrt::TransposeOp>().getPermutation();
+        if (!perm.isPermutation())
+          return failure("Transpose is not a permutation");
+        hasTransposeInput = true;
+      }
+    }
+    if (!hasTransposeInput)
+      return failure();
+
+    SmallVector<Value> newInputs;
+    for (size_t i = 0; i < op.getInputs().size(); i++) {
+      auto input = op.getInputs()[i];
+      if (auto transpose = input.getDefiningOp<tensorrt::TransposeOp>()) {
+        auto perm = transpose.getPermutation();
```

Comment:
nit: `r/auto/AffineMap`
https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable